src/compiler/checker.ts

```diff
@@ -1851,6 +1851,21 @@ namespace ts {
                         error(errorLocation, Diagnostics.Initializer_of_parameter_0_cannot_reference_identifier_1_declared_after_it, declarationNameToString(associatedDeclarationForContainingInitializer.name), declarationNameToString(<Identifier>errorLocation));
                     }
                 }
+                if (result && errorLocation && meaning & SymbolFlags.Value && !(errorLocation.flags & NodeFlags.Ambient) && isInExpressionContext(errorLocation)) {

```

Comment:
Just a reminder from the design meeting, since it looks like the new reference cache isn't the perf issue, I'd wager it's the `isInExpressionContext` call here. It might be better to annotate the node with a flag corresponding context flag to this boolean during parse if that's the case, just like how we do for `NodeFlags.JavaScriptFile` and `NodeFlags.JsonFile`, and `NodeFlags.Ambient`.

Reply:
I did some profiling after the meeting and it _seems_ like this is not problematic (`isInExpressionContext` had 1 ms aggregated total time in a ~5 second compilation), with the caveat that I haven’t yet been able to reliably reproduce as significant a perf difference as we’re seeing here. However, a node flag still sounds like a good idea (I was surprised one didn’t exist in the first place).